scripts/uniqify-module-names.py

```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python3
+
+import json
+import argparse
+import shutil
+import os
+import sh
+
+
+
+parser = argparse.ArgumentParser(description="")
+parser.add_argument("--top-filelist", type=str, required=True, help="Abs path to <top>.<model>.top.f")
+parser.add_argument("--mod-filelist", type=str, required=True, help="Abs path to <top>.<model>.model.f")
+parser.add_argument("--gen-collateral-path", dest="gcpath", type=str, required=True, help="Abs path to the gen-collateral directory")
+parser.add_argument("--model-hier-json", type=str, required=True, help="Path to hierarchy JSON emitted by firtool. Must include DUT as a module.")
+parser.add_argument('--dut', type=str, required=True, help='Name of the DUT module.')
+args = parser.parse_args()
+
+
+def get_filelist(filelist):
+  f = open(filelist, "r")
+  lines = f.readlines()
+  f.close()
+
+  fnames = []
+  for line in lines:
+    try:
+      fname = line.split("/")[-1].replace("\n", "")
+      fnames.append(fname)
+    except:
+      print(f"Something is wrong about this line {line}")
+
+  return fnames
+
+def update_filelist(cur_file, new_file):
+  sh.sed("-i", f"s/\b{cur_file}\b/{new_file}/", os.path.join(args.gcpath, args.mod_filelist))
+
+
+def generate_copy(c, sfx):
+  (cur_name, ext) = os.path.splitext(c)
+  new_name = cur_name + "_" + sfx
+  new_file = new_name + ext
+
+  cur_file = os.path.join(args.gcpath, c)
+  new_file = os.path.join(args.gcpath, new_file)
+
+  shutil.copy(cur_file, new_file)
+  sh.sed("-i", f"s/\b{cur_name}\b/{new_name}/", new_file)
+  return new_file
+
+
+
+def dfs_update_modules(tree, common_fnames, visited, top_fnames, updated_modules):
+  # List of direct submodules to update
+  childs_to_update = list()
+  for child in tree['instances']:
+    # We don't have to change stuff that are under the dut
+    if (child['module_name'] == args.dut) or (child['module_name'] in visited):
+      continue
+    if dfs_update_modules(child, common_fnames, visited, top_fnames, updated_modules):
+      childs_to_update.append(child['module_name'])
+      if (child['module_name'] + ".sv") in common_fnames:
+        child['module_name'] = child['module_name'] + "_Model"
+        updated_modules.append(child['module_name'])
+
+  cur_module = tree['module_name']
+  cur_file = cur_module + ".sv"
+  new_file = None
+
+  # cur_file is in the common list, generate a new file
+  if cur_file in common_fnames:
```

Comment:
If the `cur_file` was updated with the `_Model` suffix, wouldn't this never match since the module names originally won't have `_Model` in them (from the `common_fnames`)?

Reply:
Correct. In that case it means that the file is already updated & we don't need to bother.